clients/wallet/src/task.rs

```diff
@@ -0,0 +1,206 @@
+use crate::Ledger;
+use tokio::sync::oneshot::{channel, error::TryRecvError, Receiver, Sender};
+
+pub const LEDGER_TASK_MAX_RETRIES: u8 = 3;
+/// Determines the status of the task of processing a transaction
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum LedgerTaskStatus {
+	NotStarted,
+	/// Process Done and it was a success
+	ProcessSuccess,
+	/// The error is acceptable, and is best to just retry again
+	RecoverableError,
+	/// Something happened when
+	Error(String),
+}
+
+pub struct LedgerTask {
+	ledger: Ledger,
+
+	/// receives the status of the task execution
+	task_execution_receiver: Receiver<LedgerTaskStatus>,
+
+	/// the number of retries allowed to process this specific task
+	retries_remaining: u8,
+
+	/// The latest status of this task
+	latest_status: LedgerTaskStatus,
+}
+
+impl LedgerTask {
+	/// Creates a task structure the `TransactionEnvelope`
+	pub fn create(ledger: Ledger, task_execution_receiver: Receiver<LedgerTaskStatus>) -> Self {
+		LedgerTask {
+			ledger,
+			task_execution_receiver,
+			// The decrement action is called everytime a new receiver is set.
+			retries_remaining: LEDGER_TASK_MAX_RETRIES,
+			latest_status: LedgerTaskStatus::NotStarted,
+		}
+	}
+
+	/// Sets a new one shot receiver.
+	/// Returns true if it was successful.
+	pub fn change_receiver(&mut self, receiver: Receiver<LedgerTaskStatus>) -> bool {
```

Comment:
Maybe change the name to something like `try_recover_with_new_receiver()` or something else. With the current name, it's not obvious why I would call this function.

Reply:
updated to `fn set_receiver()` and **_private_**; only the method `fn recover_with_new_sender()` calls it.